### ISSUE TITLE
[TFLite] Potential out-of-bound array access in LOG_SOFTMAX optimized kernel

### ISSUE THREAD

**Issue by Tessil** [tensorflow/tensorflow#48521]
Hello,

The optimized LOG_SOFTMAX operator seems to have a bug that may lead to out-of-bound memory accesses.

The following [access](https://github.com/tensorflow/tensorflow/blob/89133a6a74da4b976a585b66a09a6601fce3b217/tensorflow/lite/kernels/internal/optimized/optimized_ops.h#L4143) is equivalent to `params.table[input_data[j] + max_q8 - max_val]` which can be be negative with int8 input_data.

Example:
```
input_data = [-109, -9, 0, 12, 78]
max_q8 = 127
max_val = 78
```

Access for `j = 0` is `params.table[-109 + 127 - 78]` which result in a `params.table[-60]` access.

The tests currently don't highlight the bug by luck as we end-up accessing other fields in `LogSoftmaxOpData` but replacing [`LogSoftmaxOpData`](https://github.com/tensorflow/tensorflow/blob/89133a6a74da4b976a585b66a09a6601fce3b217/tensorflow/lite/kernels/activations.cc#L87) to remove all the unused parameters (or even just putting SoftmaxParams as fist member):
```c++
struct LogSoftmaxOpData {
    struct SoftmaxParams params = {};
    float f_table[256];
};
```

makes the tests fail when running the tests in `bazel test -c opt` mode (though it may vary as it's undefined). 

Valgrind output on the test binary:
```
[ RUN      ] LogSoftmaxOpTest/LogSoftmaxOpTest.LogSoftmaxInt8/0
==8559== Invalid read of size 4
==8559==    at 0x697D124: void tflite::optimized_ops::LogSoftmax<signed char>(tflite::SoftmaxParams const&, float, tflite::RuntimeShape const&, signed char const*, tflite::RuntimeShape const&, signed char*) (optimized_ops.h:4143)
==8559==    by 0x6976EF9: TfLiteStatus tflite::ops::builtin::activations::LogSoftmaxEval<(tflite::ops::builtin::activations::KernelType)1>(TfLiteContext*, TfLiteNode*) (activations.cc:1250)
==8559==    by 0x13626C77: tflite::Subgraph::OpInvoke(TfLiteRegistration const&, TfLiteNode*) (subgraph.h:430)
==8559==    by 0x13623ED3: tflite::Subgraph::Invoke() (subgraph.cc:1062)
...
```

Thibaut

**Comment by abattery** [comment]
@jianlijianli @renjie-liu @thaink could you take a look?

**Comment by renjie-liu** [comment]
Hi Tessil, see here https://github.com/tensorflow/tensorflow/blob/89133a6a74da4b976a585b66a09a6601fce3b217/tensorflow/lite/kernels/internal/optimized/optimized_ops.h#L3780

it's always `const int32_t max_uint8 = std::numeric_limits<uint8_t>::max();` right?

**Comment by Tessil** [comment]
Hi,

It seems you're looking at the wrong kernel (Softmax instead of LogSoftmax).

The line is 
https://github.com/tensorflow/tensorflow/blob/de6d871ce4214e2856d7129deef579f2d64c8cc3/tensorflow/lite/kernels/internal/optimized/optimized_ops.h#L4137

where `T` will be `int8_t` for the int8 version of the kernel.

**Comment by renjie-liu** [comment]
Oh Thanks, will send out a fix later

**Comment by renjie-liu** [comment]
Fixed at https://github.com/tensorflow/tensorflow/commit/40fa5028dcbd9f3a70afdfc7ed98732f69235040

**Comment by Tessil** [comment]
Perfect, thank you very much. I will close the ticket then.
